### PR TITLE
[SPARK-25260][SQL] Fix namespace handling in SchemaConverters.toAvroType

### PR DESCRIPTION
**Pull request by arunmahadevan**
## What changes were proposed in this pull request?

`toAvroType` converts spark data type to avro schema. It always appends the record name to namespace so its impossible to have an Avro namespace independent of the record name.

 
When invoked with a spark data type like,

```java
val sparkSchema = StructType(Seq(
    StructField("name", StringType, nullable = false),
    StructField("address", StructType(Seq(
        StructField("city", StringType, nullable = false),
        StructField("state", StringType, nullable = false))),
    nullable = false)))
 
// map it to an avro schema with record name "employee" and top level namespace "foo.bar",
val avroSchema = SchemaConverters.toAvroType(sparkSchema,  false, "employee", "foo.bar")

// result is
// avroSchema.getName = employee
// avroSchema.getNamespace = foo.bar.employee
// avroSchema.getFullname = foo.bar.employee.employee
```
The patch proposes to fix this so that the result is

```
avroSchema.getName = employee
avroSchema.getNamespace = foo.bar
avroSchema.getFullname = foo.bar.employee
```
## How was this patch tested?

New and existing unit tests.

Please review http://spark.apache.org/contributing.html before opening a pull request.

### CODE REVIEW

**Comment by arunmahadevan** [comment]
cc @gengliangwang @dongjoon-hyun

File: external/avro/src/main/scala/org/apache/spark/sql/avro/SchemaConverters.scala [hunk 1]
```diff
@@ -143,29 +143,25 @@ object SchemaConverters {
...
-          .values(toAvroType(vt, valueContainsNull, recordName, prevNameSpace))
+          .values(toAvroType(vt, valueContainsNull, recordName, nameSpace))
       case st: StructType =>
-        val nameSpace = prevNameSpace match {
-          case "" => recordName
-          case _ => s"$prevNameSpace.$recordName"
-        }
-
+        val childNameSpace = if (nameSpace != "") s"$nameSpace.$recordName" else recordName
         val fieldsAssembler = builder.record(recordName).namespace(nameSpace).fields()
```
Comment (gengliangwang) [inline external/avro/src/main/scala/org/apache/spark/sql/avro/SchemaConverters.scala]: +1, this line is the only difference for the whole code change. The namespace here should not be the one with `recordName` at the end.

File: external/avro/src/test/scala/org/apache/spark/sql/avro/AvroSuite.scala [hunk 2]
```diff
@@ -1099,6 +1098,27 @@ class AvroSuite extends QueryTest with SharedSQLContext with SQLTestUtils {
     }
   }
 
+  test("check namespace - toAvroType") {
```
Comment (HyukjinKwon) [inline external/avro/src/test/scala/org/apache/spark/sql/avro/AvroSuite.scala]: @arunmahadevan, can we add a simple end-to-end test as well?

File: external/avro/src/test/scala/org/apache/spark/sql/avro/AvroSuite.scala [hunk 3]
```diff
@@ -1099,6 +1098,27 @@ class AvroSuite extends QueryTest with SharedSQLContext with SQLTestUtils {
...
+    val sparkSchema = StructType(Seq(
+      StructField("name", StringType, nullable = false),
+      StructField("address", StructType(Seq(
+        StructField("city", StringType, nullable = false),
+        StructField("state", StringType, nullable = false))),
+        nullable = false)))
+    val employeeType = SchemaConverters.toAvroType(sparkSchema,
+      recordName = "employee",
+      nameSpace = "foo.bar")
```
Comment (gengliangwang) [inline external/avro/src/test/scala/org/apache/spark/sql/avro/AvroSuite.scala]: nit: could you also add a case for `nameSpace` as `""` ?

**Review by gengliangwang — APPROVED** [review]
LGTM, thanks for the fix!

File: external/avro/src/test/scala/org/apache/spark/sql/avro/AvroSuite.scala (hunk 2 above)
Comment (arunmahadevan) [reply external/avro/src/test/scala/org/apache/spark/sql/avro/AvroSuite.scala]: Its sort of covered in the below existing cases. Do you think we need more?

[Validate namespace in avro file that has nested records with the same name](https://github.com/apache/spark/blob/master/external/avro/src/test/scala/org/apache/spark/sql/avro/AvroSuite.scala#L1078)
[conversion to avro and back with namespace](https://github.com/apache/spark/blob/master/external/avro/src/test/scala/org/apache/spark/sql/avro/AvroSuite.scala#L510)

File: external/avro/src/test/scala/org/apache/spark/sql/avro/AvroSuite.scala (hunk 3 above)
Comment (arunmahadevan) [reply external/avro/src/test/scala/org/apache/spark/sql/avro/AvroSuite.scala]: Added a test case for toAvroType with empty namespace

**Pull request closed without merging**